collectors/proc.plugin/proc_pagetypeinfo.c

```diff
@@ -0,0 +1,344 @@
+
+        // Pass 2: Scan the file again, with details
+        p = 0;
+        for (l=4; l < lines; l++) {
+
+            if (!pagetypeinfo_line_valid(ff, l))
+                continue;
+
+            size_t nodenum = strtoul(procfile_lineword(ff, l, 1), NULL, 10);
+            char *zonename = procfile_lineword(ff, l, 3);
+            char *typename = procfile_lineword(ff, l, 5);
+
+            // We changed node or zone
+            if (nodenum != nodenumlast || !zonenamelast ||  strncmp(zonename, zonenamelast, 6) != 0) {
+                cnt_zones++;
+                zonenamelast = zonename;
+            }
+
+            // Count the number of numa cnt_nodes
+            if( nodenum != nodenumlast ) {
+                cnt_nodes++;
+                nodenumlast = nodenum;
+            }
+
+            // Unmovable is always the first in the enum. The first line higher than 4
+            if (strncmp(typename, "Unmovable", 10) == 0 && l >4) {
+                if (cnt_pagetypes == -1)
+                    cnt_pagetypes = l - 4;
+            }
+
+            // Populate the line
+            pgl = &pagelines[p];
+
+            pgl->line = l;
+            pgl->node = nodenum;
+            pgl->type = typename;
+            pgl->zone = zonename;
+            for (o = 0; o < MAX_PAGETYPE_ORDER; o++)
+                pgl->free_pages_size[o] = str2uint64_t(procfile_lineword(ff, l, o+6)) * 1 << o;
+
+            p++;
+        }
+
+        // Init the RRD graphs
+
+        // Per-Order: sum of all node, zone, type Grouped by order
+        if (do_global == CONFIG_BOOLEAN_YES) {
+            st_order = rrdset_create_localhost(
+                "mem"
+                , "pagetype_global"
+                , NULL
+                , "pagetype"
+                , NULL
+                , "System orders available"
+                , "B"
+                , PLUGIN_PROC_NAME
+                , PLUGIN_PROC_MODULE_PAGETYPEINFO_NAME
+                , NETDATA_CHART_PRIO_SYSTEM_MEMFRAG
+                , update_every
+                , RRDSET_TYPE_STACKED
+            );
+            for (o = 0; o < MAX_PAGETYPE_ORDER; o++) {
+                char id[3+1];
+                snprintfz(id, 3, "%lu", o);
+
+                char name[20+1];
+                dim_name(name, o);
+
+                systemorders[o].rd = rrddim_add(st_order, id, name, PAGE_SIZE, 1, RRD_ALGORITHM_ABSOLUTE);
+                rrddim_set_by_pointer(st_order, systemorders[o].rd, systemorders[o].size);
+            }
+        }
+
+
+        // Per-Numa Node & Zone & Type (full detail). Only if sum(line) > 0
+        st_nodezonetype = callocz(cnt_zones, sizeof(RRDSET));
+        for (p = 0; p < pagelines_cnt; p++) {
+            pgl = &pagelines[p];
+
+            // Skip invalid, refused or empty pagelines if not explicitely requested
+            if (!pgl
+                || do_detail == CONFIG_BOOLEAN_NO
+                || (do_detail == CONFIG_BOOLEAN_AUTO && pageline_total_count(pgl) == 0))
```

Comment:
Can't pages be moved to other types at runtime?

Reply:
They may, but I'm not sure how to make it "dynamic". Most of the time, only 3 types over the 6 are filled, and putting them all would generate 3 not-shown-types * (NUMA-Nodes + 2 DMA/DMA32 charts, each with 10 dimensions, and all at 0.

So if config `enable detail per-type` is forced to `yes`, it'll show them all, as user explicitely wants them

else, the default `auto` will only show non-empty lines at discovery time. i'm no sure i should rescan often, nor we should create ephemeral graph, as I had issues with the cleanup of and processing of ephemeral cgroups

Reply:
You can initialize charts and dimensions right before setting the dimensions like in
https://github.com/netdata/netdata/blob/master/collectors/proc.plugin/proc_meminfo.c#L222-L257

You can also make them fully dynamic using `rrdset_is_obsolete()` and `rrddim_is_obsolete()` but it doesn't make sense if data changes are not generally expected.

Reply:
Please pay attention to `netdata_zero_metrics_enabled == CONFIG_BOOLEAN_YES`. It is a global option to enable `auto` charts permanently.

Reply:
I've added the configuration check for `netdata_zero_metrics_enabled`.
I'm still checking for fully dynamic rrd.

Reply:
Do you want to make charts more dynamic here or in a separate PR?

Reply:
On a separate PR would be better, depending on user feedback.